llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1613,6 +1613,22 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
   if (Instruction *Overflow = foldLShrOverflowBit(I))
     return Overflow;
 
+  // Transform ((pow2 << x) >> cttz(pow2 << y)) -> ((1 << x) >> y)
+  Value *Shl0_Op0, *Shl0_Op1, *Shl1_Op0, *Shl1_Op1;
+  BinaryOperator *Shl1;
+  if (match(Op0, m_Shl(m_Value(Shl0_Op0), m_Value(Shl0_Op1))) &&
+      match(Op1, m_Intrinsic<Intrinsic::cttz>(m_BinOp(Shl1))) &&
+      match(Shl1, m_Shl(m_Value(Shl1_Op0), m_Value(Shl1_Op1))) &&
+      isKnownToBeAPowerOfTwo(Shl1, false, 0, SQ.getWithInstruction(&I).CxtI) &&
+      Shl0_Op0 == Shl1_Op0) {
+    auto *Shl0 = cast<BinaryOperator>(Op0);
+    if ((Shl0->hasNoUnsignedWrap() && Shl1->hasNoUnsignedWrap()) ||
+        (Shl0->hasNoSignedWrap() && Shl1->hasNoSignedWrap())) {
```

Comment:
Proof for nsw?

Reply:
Sorry if I've misunderstood your question,
If you're asking for alive test I think your comment from earlier proves it: https://alive2.llvm.org/ce/z/gWbtPw
If you're asking for an IR test, InstCombine sets nsw/nuw flags very early on for shift binops, so a separate IR test for both nuw and nsw are actually testing both nsw and nuw

Reply:
> If you're asking for an alive test I think your comment from earlier proves it: https://alive2.llvm.org/ce/z/gWbtPw

Please add it to the PR description.

Reply:
Done